Iewpf interpolation

havahol opened the pull request:
Updates the IEWPF algorithm to include the interpolation operator.
Fixes the unit tests for IEWPF
Explores each step of the IEWPF algorithm in the IEWPF_GPU.ipynb notebook.

Relevant unittests (in `gpu_ocean/tests/`):
`python random_numbers.py 1`
`python particle_filter_tests.py 5`

Relevant notebooks in `gpu_ocean/prototypes`:
`IEWPF_GPU.ipynb`

This PR also contains the review requests from PR #148

havahol commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement interpolation operator within the model error covariance structure, so that the correlation radius is no longer tightly related to the grid resolution.](https://trello.com/c/vPfkOPH0/96-implement-interpolation-operator-within-the-model-error-covariance-structure-so-that-the-correlation-radius-is-no-longer-tightly)

setmar commented on gpu_ocean/SWESimulators/BaseOceanStateEnsemble.py:
```diff
@@ -85,10 +114,18 @@ def cleanUp(self):
     def setGridInfo(self, nx, ny, dx, dy, dt, 
                     boundaryConditions=Common.BoundaryConditions(), 
                     eta=None, hu=None, hv=None, H=None):
+        warnings.warn('The function setGridInfo will be deprecated. Please use the improved constructor instead',
```
This should probably be "is deprecated" or "will be removed". Marking functions for later removal _is_ deprecation. Goes for all instances of DeprecationWarning in the PR.

setmar commented on gpu_ocean/SWESimulators/BaseOceanStateEnsemble.py:
```diff
@@ -954,6 +1028,8 @@ def plotVelocityInfo(self, title=None, printInfo=False):
 
         # PLOT POSITIONS OF PARTICLES AND OBSERVATIONS
+
+        #ax = plt.subplot2grid((plotRows,3), (0,0), polar=True, axisbg='#ffffff')
```
Unused code.

setmar commented on gpu_ocean/SWESimulators/BaseOceanStateEnsemble.py:
```diff
@@ -996,6 +1072,7 @@ def plotVelocityInfo(self, title=None, printInfo=False):
 
         if self.driftersPerOceanModel > 1:
             for drifter_id in range(0,min(3, self.driftersPerOceanModel)):
+                #ax = plt.subplot2grid((plotRows,3), (2,drifter_id-1), polar=True, axisbg='#ffffff')
```
Unused code.

setmar commented on gpu_ocean/SWESimulators/IEWPFOcean.py:
```diff
@@ -554,25 +614,69 @@ def _createS(self, ensemble):
+            # north and east values are calculated from the surface in the north-east direction
+            # south and west values are calculated from the surface in the south-west direction 
+            
+            x_north = interpolation_alias(x_corr, mid_i  , mid_j  ,          0.0,      rel_offset)
```
Please add whitespace only after commas, not before :)

(Will not comment this in the rest of the code.)

havahol replied:
As I see it, the only reason to use more than one space in places like this, is to align parameters across multiple lines. For me, it is then also natural to align the commas, and I don't really see why that should be a problem :)

setmar replied:
Parameters can be aligned with whitespace after commas (or other punctuation). I think whitespace _before_ commas makes the code harder to read. I also think this is supported in most standardized code styles. I will merge this PR, but please consider not using whitespace _before_ commas :)

setmar commented on gpu_ocean/SWESimulators/OceanStateNoise.py:
```diff
@@ -617,82 +666,65 @@ def _interpolate_CPU(self, coarse_eta, interpolation_order=3):
-                fxy01 = (fy11 - fy_11)/2
-                fxy10 = (fy20 -  fy00)/2
-                fxy11 = (fy21 -  fy01)/2
+        #print("(min_rel_x, max_rel_x)", (min_rel_x, max_rel_x))
```
Unused code.

havahol commented:
Removed the unused code.

Removed the deprecated functions since they no longer were reachable by old code either. Went through the SEnKF notebook to confirm this, so made it compatible with py3. It does not need testing.
